File: internal/hooks/payload.go
package hooks

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"time"
)

// EventType enumerates the cluster lifecycle events that can trigger webhooks.
type EventType string

const (
	EventTaskCreated   EventType = "task_created"
	EventTaskCompleted EventType = "task_completed"
	EventTaskFailed    EventType = "task_failed"
	EventNodeJoined    EventType = "node_joined"
	EventNodeLeft      EventType = "node_left"
	EventLeaseCreated  EventType = "lease_created"
	EventLeaseExpired  EventType = "lease_expired"
)

// AllEventTypes returns the complete list of supported event types.
func AllEventTypes() []EventType {
	return []EventType{
		EventTaskCreated,
		EventTaskCompleted,
		EventTaskFailed,
		EventNodeJoined,
		EventNodeLeft,
		EventLeaseCreated,
		EventLeaseExpired,
	}
}

// IsValidEvent returns true if the event type is a known cluster event.
func IsValidEvent(et EventType) bool {
	for _, e := range AllEventTypes() {
		if e == et {
			return true
		}
	}
	return false
}

// Payload is the JSON body sent to webhook endpoints.
type Payload struct {
	EventType EventType   `json:"event_type"`
	Timestamp time.Time   `json:"timestamp"`
	Data      interface{} `json:"data"`
}

// DeliveryStatus records the outcome of a single delivery attempt.
type DeliveryStatus string

const (
	DeliverySuccess    DeliveryStatus = "success"
	DeliveryFailed     DeliveryStatus = "failed"
	DeliveryPending    DeliveryStatus = "pending"
	DeliveryRetry      DeliveryStatus = "retry"
)

// Delivery represents one attempt to deliver a webhook payload.
type Delivery struct {
	ID          string         `json:"id"`
	HookID      string         `json:"hook_id"`
	EventType   EventType      `json:"event_type"`
	URL         string         `json:"url"`
	Status      DeliveryStatus `json:"status"`
	StatusCode  int            `json:"status_code,omitempty"`
	Error       string         `json:"error,omitempty"`
	Attempts    int            `json:"attempts"`
	MaxAttempts int            `json:"max_attempts"`
	CreatedAt   time.Time      `json:"created_at"`
	UpdatedAt   time.Time      `json:"updated_at"`
}

// SignPayload computes HMAC-SHA256 over the raw body bytes using the given secret.
// Returns the hex-encoded signature string suitable for the X-Hub-Signature-256 header.
func SignPayload(body []byte, secret string) string {
	mac := hmac.New(sha256.New, []byte(secret))
	mac.Write(body)
	return fmt.Sprintf("sha256=%s", hex.EncodeToString(mac.Sum(nil)))
}

// VerifySignature verifies the HMAC-SHA256 signature of raw body bytes.
func VerifySignature(body []byte, secret, signature string) bool {
	expected := SignPayload(body, secret)
	return hmac.Equal([]byte(expected), []byte(signature))
}

File: hermes_cluster/hooks/payload.py
"""Webhook event types, payloads, and HMAC signing (from hooks/payload.go)."""

from __future__ import annotations

import hashlib
import hmac
from datetime import datetime, timezone
from enum import Enum
from typing import Any

from pydantic import BaseModel, Field


class HookEventType(str, Enum):
    """Cluster lifecycle events that trigger webhooks."""
    TASK_CREATED = "task_created"
    TASK_COMPLETED = "task_completed"
    TASK_FAILED = "task_failed"
    NODE_JOINED = "node_joined"
    NODE_LEFT = "node_left"
    LEASE_CREATED = "lease_created"
    LEASE_EXPIRED = "lease_expired"


# All valid event types for quick lookup
_ALL_EVENTS: frozenset[HookEventType] = frozenset(HookEventType)


def all_event_types() -> list[HookEventType]:
    """Return complete list of supported event types."""
    return list(HookEventType)


def is_valid_event(event_type: str) -> bool:
    """Check if a string is a valid event type."""
    return event_type in _ALL_EVENTS


class DeliveryStatus(str, Enum):
    """Webhook delivery states."""
    SUCCESS = "success"
    FAILED = "failed"
    PENDING = "pending"
    RETRY = "retry"


class Payload(BaseModel):
    """JSON body sent to webhook endpoints."""
    event_type: HookEventType = Field(default=HookEventType.TASK_CREATED, alias="event_type")
    timestamp: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    data: Any = None


def sign_payload(body: bytes, secret: str) -> str:
    """Compute HMAC-SHA256 over raw body bytes.

    Returns hex-encoded signature string suitable for X-Hub-Signature-256 header.
    Format: sha256=<hex>
    """
    mac = hmac.new(secret.encode("utf-8"), body, hashlib.sha256)
    return f"sha256={mac.hexdigest()}"


def verify_signature(body: bytes, secret: str, signature: str) -> bool:
    """Verify HMAC-SHA256 signature of raw body bytes.

    Uses constant-time comparison to prevent timing attacks.
    """
    expected = sign_payload(body, secret)
    return hmac.compare_digest(expected, signature)
